### PR TITLE
Avoid autocast deprecation warning in DataParallel

### LINKED ISSUE #130659: Internal uses of `torch.cuda.amp.autocast` raise FutureWarnings
### 🐛 Describe the bug

PyTorch 2.4 deprecated the use of `torch.cuda.amp.autocast` in favor of `torch.amp.autocast("cuda", ...)`, but this change has missed updating internal uses in PyTorch. For example in DP here:
https://github.com/pytorch/pytorch/blob/3710a7962261f26bf9ad8b768229582bf9f00c94/torch/nn/parallel/parallel_apply.py#L92-L93

```py
import torch

model = torch.nn.Linear(10, 10).cuda()
model = torch.nn.DataParallel(model, device_ids=[0, 1])
output = model(torch.randn(20, 10).cuda())
```
Produces:
```
/home/adrian/.conda/envs/lightning/lib/python3.10/site-packages/torch/nn/parallel/parallel_apply.py:79: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with torch.cuda.device(device), torch.cuda.stream(stream), autocast(enabled=autocast_enabled):
/home/adrian/.conda/envs/lightning/lib/python3.10/site-packages/torch/nn/modules/linear.py:117: UserWarning: Attempting to run cuBLAS, but there was no current CUDA context! Attempting to set the primary context... (Triggered internally at ../aten/src/ATen/cuda/CublasHandlePool.cpp:135.)
  return F.linear(input, self.weight, self.bias)
```

Since these are caused internally, the user will see the warning but not be able to do anything with it. 

If desired, I can send a PR with a fix for this :)

### Versions

PyTorch version: 2.4.0+cu121

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

### PR DESCRIPTION
**Pull request by awaelchli**
Fixes #130659

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/nn/parallel/parallel_apply.py [hunk 1]
```diff
@@ -89,8 +88,8 @@ def _worker(
         if stream is None:
             stream = torch.cuda.current_stream(device)
         try:
-            with torch.cuda.device(device), torch.cuda.stream(stream), autocast(
-                enabled=autocast_enabled
+            with torch.cuda.device(device), torch.cuda.stream(stream), torch.amp.autocast(
+                "cuda", enabled=autocast_enabled
```
Comment (guangyey) [inline torch/nn/parallel/parallel_apply.py:92]:
```suggestion
            with torch.cuda.device(device), torch.cuda.stream(
                stream
            ), torch.amp.autocast("cuda", enabled=autocast_enabled):
```

**Review by guangyey — APPROVED** [review]

File: torch/nn/parallel/parallel_apply.py [hunk 2]
```diff
@@ -89,8 +88,9 @@ def _worker(
         if stream is None:
             stream = torch.cuda.current_stream(device)
         try:
-            with torch.cuda.device(device), torch.cuda.stream(stream), autocast(
-                enabled=autocast_enabled
+            with torch.cuda.device(device), torch.cuda.stream(
+                stream
+            ), torch.amp.autocast("cuda", enabled=autocast_enabled):
             ):
```
Comment (awaelchli) [inline torch/nn/parallel/parallel_apply.py:94]:
```suggestion
```

**Review by fegin — APPROVED** [review]

**Review by albanD — APPROVED** [review]

**Comment by awaelchli** [comment]
Some test workflows failed, but look unrelated. How to proceed?

**Comment by guangyey** [comment]
Unrelated failures.
@pytorchbot merge -i

**Comment by guangyey** [comment]
Unrelated failures.
@pytorchbot merge -i

**Pull request closed without merging**